tools/reduce.py

```diff
@@ -0,0 +1,303 @@
+    # path_flag_pattern: re.Pattern
+
+    def __post_init__(self):
+        # Lenient behavior instead of hard-coding which flags allow =
+        flags = [flag + "=?" for flag in PATH_FLAGS]
+        self.path_flag_pattern = re.compile(
+            "(?P<flag>{})(?P<path>.*)".format("|".join(flags))
+        )
+
+    def run(self, start_tu_path: pathlib.Path):
+        assert start_tu_path.is_absolute()
+        absolute_start_tu_path = start_tu_path
+        start_tu_path = start_tu_path.relative_to(self.project_root)
+        assert not start_tu_path.is_absolute()
+
+        original_tu_path = str(self.entry.filepath)
+        for i, arg in enumerate(self.entry.arguments):
+            if original_tu_path in arg:
+                self.entry.arguments[i] = str(start_tu_path)
+        self.entry.filepath = start_tu_path
+
+        self.fix_paths()
+
+        with contextlib.ExitStack() as exitstack:
+            tempdir = tempfile.mkdtemp(prefix="scip-clang-reduce-")
+            exitstack.callback(
+                lambda: logging.info(
+                    f"scip-clang temporary directory persisted at {tempdir}"
+                )
+            )
+            tempdir = pathlib.Path(tempdir)
+            self.entry.directory = tempdir
+
+            shutil.copyfile(absolute_start_tu_path, tempdir.joinpath(start_tu_path))
+            compdb = tempdir.joinpath("compile_commands.json")
+            with open(compdb, "w") as compdb_file:
+                json.dump([self.entry.to_dict()], compdb_file)
+
+            run_sh = tempdir.joinpath("run.sh")
+            with open(run_sh, "w") as run_sh_file:
+                assert self.scip_clang.is_absolute()
+                run_sh_file.write(
+                    textwrap.dedent(
+                        f"""\
+                    #!/usr/bin/env bash
+                    {self.scip_clang} --worker-mode=compdb --compdb-path={compdb} > out.log 2>&1
+                    set -e
+                    grep -E '{self.scip_clang_output_pattern.pattern}' out.log
+                    """
+                    )
+                )
+            run_sh.chmod(run_sh.stat().st_mode | stat.S_IEXEC)
+
+            result = subprocess.run(
+                ["creduce", str(run_sh), str(start_tu_path)] + self.extra_args,
+                cwd=tempdir,
+            )
+            if result.returncode != 0:
+                sys.exit(result.returncode)
+
+    def fix_paths(self):
+        """
+        Adjusts the paths in the argument list so that the compilation database
+        entry can be relocated correctly to a temporary directory.
+
+        Post-condition:
+        - Include paths, such as using -I and -isysroot use absolute paths
+        - The path for the main TU has a relative path
+
+        Creduce creates temporary directories, so we want all the paths to
+        work properly from those directories too.
+        """
+        original_path = self.entry.filepath
+        if self.entry.filepath.is_absolute():
```

Comment:
...if the filepath is relative, are we on the road to ruin? Bit confused at this point. A relative path is OK because we're going to copy a bunch of stuff into the temp directory such that the relative paths stay valid?

Reply:
> A relative path is OK because we're going to copy a bunch of stuff into the temp directory such that the relative paths stay valid?

Yep, exactly. We want to make sure any mentions of the main TU path are relative, since C-reduce copies files into temporary directories to run commands.